<a href="https://colab.research.google.com/github/Ridhof177/Tugas-Mata-Kuliah-Teknik-Pengambilan-Sampel-dan-Data-Wrangling./blob/main/TUGAS_2_TPSDW_RIDHO_FATILLA_WIBAWA_SAINS_DATA_24_241061010.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TUGAS 2 - TEKNIK PENGAMBILAN SAMPEL DAN DATA WRANGLING
## Mahasiswa diminta melakukan tahapan data preparation menggunakan Python (Pandas dan Numpy) atau tools analisis data lainnya dengan langkah-langkah berikut.

NAMA  : RIDHO FATILLA WIBAWA

NIM   : 241061010

PRODI : SAINS DATA 24

### 1. Menampilkan Jumlah Data Hilang

Lakukan pemeriksaan terhadap dataset untuk mengetahui jumlah data hilang (missing values) pada setiap variabel.

Instruksi:

•	Tampilkan jumlah missing value pada setiap kolom.

•	Jelaskan variabel mana yang memiliki data hilang paling banyak.


In [4]:
# Import library
import pandas as pd
import numpy as np

# Membaca file dataset csv
sd = pd.read_csv('/content/data_tugas 2.csv')

# Menampilkan 7 data teratas
sd.head(7)

,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height
0,?,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
1,?,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
2,?,alfa-romero,gas,two,hatchback,94.5,171.2,65.5,52.4
3,164,audi,gas,four,sedan,99.8,176.6,66.2,54.3
4,164,audi,gas,four,sedan,99.4,176.6,66.4,54.3
5,?,audi,gas,two,sedan,99.8,177.3,66.3,53.1
6,158,audi,gas,four,sedan,105.8,192.7,71.4,55.7


Pada langkah ini dilakukan proses awal dalam data preparation, yaitu mengimpor library, membaca dataset, dan menampilkan sebagian data.

- Pandas digunakan untuk mengolah data dalam bentuk tabel (DataFrame), seperti membaca file dan menampilkan data.
- NumPy digunakan untuk operasi numerik serta membantu dalam penanganan nilai kosong seperti NaN.

Dataset dibaca menggunakan fungsi read_csv() dan disimpan dalam variabel sd. Selanjutnya, ditampilkan 7 baris pertama menggunakan head(7) untuk melihat struktur data.

Dari hasil tampilan, dapat ditemukan adanya nilai "?" pada beberapa variabel. Nilai tersebut menunjukkan data yang hilang, namun belum dikenali sebagai missing value oleh pandas. Oleh karena itu, nilai "?" perlu dikonversi menjadi NaN agar dapat diproses pada tahap berikutnya.

In [5]:
# Mengganti tanda '?' menjadi NaN (missing value)
sd.replace('?', np.nan, inplace=True)

# Mengecek jumlah missing value pada setiap kolom
missing_values = sd.isnull().sum()

# Menampilkan Tabel Jumlah Missing Values
print("Tabel Jumlah Missing Values Pada Setiap Kolom:")
print(missing_values)

Tabel Jumlah Missing Values Pada Setiap Kolom:
normalized-losses    41
make                  4
bahan-bakar          13
jumlah-pintu         18
body-style           12
wheel-bases           3
length                8
width                 5
height                4
dtype: int64


In [6]:
sd.head(7)

,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height
0,NaN,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
1,NaN,alfa-romero,gas,two,convertible,88.6,168.8,64.1,48.8
2,NaN,alfa-romero,gas,two,hatchback,94.5,171.2,65.5,52.4
3,164,audi,gas,four,sedan,99.8,176.6,66.2,54.3
4,164,audi,gas,four,sedan,99.4,176.6,66.4,54.3
5,NaN,audi,gas,two,sedan,99.8,177.3,66.3,53.1
6,158,audi,gas,four,sedan,105.8,192.7,71.4,55.7


Berdasarkan tabel jumlah missing value pada setiap kolom, dapat diketahui bahwa masih terdapat data yang hilang pada beberapa variabel dalam dataset.

**Variabel dengan jumlah missing value terbanyak adalah normalized-losses sebanyak 41 data.** Hal ini menunjukkan bahwa variabel tersebut memiliki tingkat ketidaklengkapan data yang paling tinggi dan berpotensi mempengaruhi hasil analisis jika tidak ditangani dengan baik.

Selanjutnya, variabel jumlah-pintu memiliki 18 data hilang, diikuti oleh:

- bahan-bakar sebanyak 13 data
- body-style sebanyak 12 data

Sementara itu, variabel numerik seperti:

- length (8 data)
- width (5 data)
- height (4 data)
- wheel-bases (3 data)

memiliki jumlah missing value yang relatif lebih sedikit.

bisa juga di perhatikan pada sd.head(7) tanda "?" telah di ubah menjadi NaN.

### 2. Mengganti Missing Value pada Variabel Kategori

Untuk variabel berbentuk kategori, data hilang harus diganti dengan kategori baru yaitu "unknown".

Variabel kategorik pada dataset:

•	make

•	bahan-bakar

•	jumlah-pintu

•	body-style

Instruksi:

•	Identifikasi missing value pada variabel kategori

•	Ganti nilai tersebut dengan "unknown"


In [7]:
# Daftar variabel kategorik
kategori = ['make', 'bahan-bakar', 'jumlah-pintu', 'body-style']

# Mengganti missing value dengan 'unknown'
for col in kategori:
    sd[col] = sd[col].fillna('unknown')

# Menampilkan ringkasan jumlah setiap kategori
for col in kategori:
    print(f"\nTabel kategori untuk kolom: {col}")
    print(sd[col].value_counts())


Tabel kategori untuk kolom: make
make
toyota           32
mazda            17
nissan           17
honda            13
mitsubishi       12
volkswagen       12
volvo            11
subaru           11
peugot           11
dodge             9
mercedes-benz     8
bmw               8
plymouth          7
audi              7
saab              6
porsche           4
unknown           4
isuzu             4
alfa-romero       3
chevrolet         3
jaguar            3
renault           2
mercury           1
Name: count, dtype: int64

Tabel kategori untuk kolom: bahan-bakar
bahan-bakar
gas        173
diesel      19
unknown     13
Name: count, dtype: int64

Tabel kategori untuk kolom: jumlah-pintu
jumlah-pintu
four       101
two         86
unknown     18
Name: count, dtype: int64

Tabel kategori untuk kolom: body-style
body-style
sedan          88
hatchback      67
wagon          24
unknown        12
hardtop         8
convertible     6
Name: count, dtype: int64


Tabel ringkasan menunjukkan jumlah kemunculan setiap kategori pada variabel kategorik setelah missing value diganti dengan "unknown".

- Pada variabel make, kategori "unknown" berjumlah 4 data, menunjukkan jumlah data yang sebelumnya hilang.
- Pada bahan-bakar, kategori "unknown" berjumlah 13 data, sedangkan kategori terbanyak adalah "gas" (173 data).
- Pada jumlah-pintu, kategori "unknown" berjumlah 18 data, dengan kategori utama "four" (101 data) dan "two" (86 data).
- Pada body-style, kategori "unknown" berjumlah 12 data, dengan kategori terbanyak "sedan" (88 data).

Nilai-nilai tersebut menunjukkan distribusi data pada setiap kategori serta jumlah data yang sebelumnya hilang.

### 3. Mengganti Missing Value pada Variabel Numerik

Untuk variabel numerik yang memiliki data hilang, lakukan imputasi menggunakan nilai rata-rata (mean).

Variabel numerik pada dataset:

•	length

•	width

•	height

Instruksi:

•	Hitung nilai rata-rata masing-masing variabel

•	Gunakan nilai tersebut untuk mengganti data yang hilang.


In [9]:
# Daftar variabel numerik
numerik = ['length', 'width', 'height']

# Menghitung mean dan mengganti missing value
for col in numerik:
    mean_value = sd[col].astype(float).mean()
    sd[col] = sd[col].astype(float).fillna(mean_value)
    print(f"Mean {col}: {mean_value}")


Mean length: 173.90253807106598
Mean width: 65.931
Mean height: 53.73731343283582


In [11]:
sd.head(10)

,normalized-losses,make,bahan-bakar,jumlah-pintu,body-style,wheel-bases,length,width,height
0,NaN,alfa-romero,gas,two,convertible,88.6,168.800000,64.1,48.8
1,NaN,alfa-romero,gas,two,convertible,88.6,168.800000,64.1,48.8
2,NaN,alfa-romero,gas,two,hatchback,94.5,171.200000,65.5,52.4
3,164,audi,gas,four,sedan,99.8,176.600000,66.2,54.3
4,164,audi,gas,four,sedan,99.4,176.600000,66.4,54.3
5,NaN,audi,gas,two,sedan,99.8,177.300000,66.3,53.1
6,158,audi,gas,four,sedan,105.8,192.700000,71.4,55.7
7,NaN,audi,gas,four,wagon,105.8,173.902538,71.4,55.7
8,158,audi,gas,four,sedan,105.8,192.700000,71.4,55.9
9,NaN,audi,gas,two,hatchback,99.5,178.200000,67.9,52.0


Output yang ditampilkan merupakan nilai rata-rata (mean) dari masing-masing variabel numerik, yaitu:

- length ≈ 173.90
- width ≈ 65.93
- height ≈ 53.74

Nilai mean tersebut digunakan untuk menggantikan data yang hilang pada masing-masing variabel.

Dengan menggunakan nilai rata-rata, data yang kosong dapat diisi dengan nilai yang mewakili kondisi umum dari data tersebut, sehingga tidak mengganggu distribusi data secara signifikan.

### 4. Min-Max Normalization
Lakukan Min-Max Normalization pada variabel:

wheel-bases


In [13]:
# Menghitung nilai minimum dan maksimum
min_val = sd['wheel-bases'].astype(float).min()
max_val = sd['wheel-bases'].astype(float).max()

# Melakukan Min-Max Normalization
sd['wheel-bases-normalized'] = (sd['wheel-bases'].astype(float) - min_val) / (max_val - min_val)

# Menampilkan hasil
sd[['wheel-bases', 'wheel-bases-normalized']].head()

,wheel-bases,wheel-bases-normalized
0,88.6,0.058309
1,88.6,0.058309
2,94.5,0.230321
3,99.8,0.384840
4,99.4,0.373178


Pada langkah ini dilakukan normalisasi data menggunakan metode Min-Max Normalization pada variabel wheel-bases.

Metode ini mengubah nilai data ke dalam rentang 0 hingga 1:
- Rumus Min-Max Normalization

$$
X' = \frac{X - X_{min}}{X_{max} - X_{min}}
$$

- min() digunakan untuk mencari nilai minimum
- max() digunakan untuk mencari nilai maksimum
- Hasil normalisasi disimpan dalam kolom baru yaitu wheel-bases-normalized

Nilai pada kolom wheel-bases-normalized merupakan hasil transformasi dari nilai asli menggunakan metode Min-Max Normalization.

- Nilai 88.6 diubah menjadi 0.058309

- Nilai 94.5 diubah menjadi 0.230321

- Nilai 99.8 diubah menjadi 0.384840

Perubahan ini terjadi karena setiap nilai disesuaikan berdasarkan nilai minimum dan maksimum dari seluruh data.
Hasil output menunjukkan bahwa variabel wheel-bases telah berhasil dinormalisasi ke dalam skala 0–1 melalui kolom baru wheel-bases-normalized, sehingga memudahkan proses analisis dan pemodelan selanjutnya.

### 5. Z-Score Normalization

Lakukan Z-score normalization pada variabel berikut:

•	length

•	width

•	height


In [14]:
# Daftar variabel numerik
numerik = ['length', 'width', 'height']

# Melakukan Z-Score Normalization
for col in numerik:
    mean = sd[col].astype(float).mean()
    std = sd[col].astype(float).std()

    sd[col + '-zscore'] = (sd[col].astype(float) - mean) / std

# Menampilkan hasil
sd[['length', 'length-zscore',
    'width', 'width-zscore',
    'height', 'height-zscore']].head()

,length,length-zscore,width,width-zscore,height,height-zscore
0,168.8,-0.428526,64.1,-0.867630,48.8,-2.038094
1,168.8,-0.428526,64.1,-0.867630,48.8,-2.038094
2,171.2,-0.226967,65.5,-0.204232,52.4,-0.552035
3,176.6,0.226541,66.2,0.127467,54.3,0.232274
4,176.6,0.226541,66.4,0.222238,54.3,0.232274


Pada langkah ini dilakukan normalisasi data menggunakan metode Z-Score pada variabel length, width, dan height.

Metode Z-Score digunakan untuk mengubah data sehingga memiliki:

- Rata-rata (mean) ≈ 0
- Standar deviasi (std) ≈ 1

### Rumus Z-Score Normalization

$$
Z = \frac{X - \mu}{\sigma}
$$


Hasil normalisasi disimpan dalam kolom baru:

- length-zscore
- width-zscore
- height-zscore

Output menampilkan nilai asli dan hasil normalisasi Z-score.

- Nilai positif → di atas rata-rata
- Nilai negatif → di bawah rata-rata
- Nilai mendekati 0 → dekat dengan rata-rata

Dengan normalisasi ini, data menjadi lebih terstandarisasi tanpa mengubah pola distribusinya.

Variabel numerik telah berhasil dinormalisasi menggunakan metode Z-Score, sehingga data memiliki skala yang seragam dan lebih optimal untuk analisis maupun pemodelan.